src/hotspot/share/opto/callGenerator.cpp

```diff
@@ -809,6 +813,424 @@ CallGenerator* CallGenerator::for_vector_reboxing_late_inline(ciMethod* method,
+#ifdef _LP64
+                assert(offset2->Opcode() == Op_ConvI2L, "");
+                offset2 = offset2->in(1);
+                if (offset2->Opcode() == Op_CastII && offset2->in(0)->is_Proj() &&
+                    offset2->in(0)->in(0) == get_cache_iff) {
+                  ShouldNotReachHere();
+                  offset2 = offset2->in(1);
+                }
+#endif
+                index = kit.gvn().transform(new AddINode(offset2, index));
+              }
+
+              if (get_first_iff == nullptr) {
+                get_first_iff = c->as_If();
+                first_index = index;
+              } else {
+                assert(get_second_iff == nullptr, "");
+                get_second_iff = c->as_If();
+                second_index = index;
+              }
+            }
+          } else if (c->is_RangeCheck()) {
+            // Kill the range checks as they are known to always succeed
+            kit.gvn().hash_delete(c);
+            c->set_req(1, kit.gvn().intcon(1));
+            C->record_for_igvn(c);
+          } else if (c->is_CallStaticJava()) {
+            assert(slow_call == nullptr, "");
+            slow_call = c->as_CallStaticJava();
+            assert(slow_call->method()->intrinsic_id() == vmIntrinsics::_SVslowGet, "");
+          } else {
+            assert(c->is_Proj() || c->is_Catch(), "");
+          }
+          wq.push(c->in(0));
+        }
+      }
+      // get_first_iff/get_second_iff contain the first/second check we ran into during the graph traversal but they may
+      // not be the first/second one in execution order. Perform another traversal to figure out which is first.
+      if (get_second_iff != nullptr) {
+        Node_Stack stack(0);
+        stack.push(get_cache_iff, 0);
+        while (stack.is_nonempty()) {
+          Node* c = stack.node();
+          uint i = stack.index();
+          if (i < c->outcnt()) {
+            stack.set_index(i + 1);
+            Node* u = c->raw_out(i);
+            if (wq.member(u) && u != c) {
+              if (u == get_first_iff) {
+                break;
+              } else if (u == get_second_iff) {
+                swap(get_first_iff, get_second_iff);
+                swap(first_index, second_index);
+                break;
+              }
+              stack.push(u, 0);
+            }
+          } else {
+            stack.pop();
+          }
+        }
+      }
+    }
+
+    assert(get_cache_iff != nullptr, "");
+    assert(get_second_iff == nullptr || get_first_iff != nullptr, "");
+
+    if (get_first_iff != nullptr && get_second_iff != nullptr) {
+      ProjNode* get_first_iff_failure = get_first_iff->proj_out(
+              get_first_iff->in(1)->as_Bool()->_test._test == BoolTest::ne ? 0 : 1);
+      CallStaticJavaNode* get_first_iff_unc = get_first_iff_failure->is_uncommon_trap_proj(Deoptimization::Reason_none);
+      if (get_first_iff_unc != nullptr) {
+        // first cache check never hits, keep only the second.
```

Comment:
I'm struggling to understand:
We still have an unc-trap for the first. So we never failed so far, right? So we always found it in the cache, or am I wrong?
We are not removing this unc-trap though, right?

Reply:
The `ScopedValue.get()` codee probes 2 cache locations. If, when pattern matching the `get()` subgraph:

- we only find a single if that probes the cache, then, according to profile data, there was always a hit at the first cache location.

- we find 2 ifs, then the first and second locations were probed. If the first if's other branch is to an uncommon trap, then that location never saw a cache hit. In that case, when the `ScopedValueGetHitsInCacheNode` is expanded, only code to probe the second location is added back to the IR.